src/wp-includes/html-api/class-wp-html-tag-processor.php

```diff
@@ -1155,12 +1168,7 @@ public function class_list() {
 				return;
 			}
 
-			/*
-			 * CSS class names are case-insensitive in the ASCII range.
-			 *
-			 * @see https://www.w3.org/TR/CSS2/syndata.html#x1
-			 */
-			$name = strtolower( substr( $class, $at, $length ) );
+			$name = substr( $class, $at, $length );
```

Comment:
why not go ahead and base this off of the document compatability mode?

Reply:
okay I see. well, for adding it to the `$seen` array I can see my comment being valid. lower-case the seen list so we can filter duplicates, but reports the casing of the first match as the casing of the attribute. this preserves author intent as far as we can make it while being reasonable.

Reply:
I'll state the problem for clarity. In quirks mode we can deduplicate class name case-insensitively, but then what form should be yielded? Lowercase (this was the behavior before)? Uppercase? First/last/nth occurrence in HTML? For example, all the class names in this HTML are identical CSS class names, so if _one_ is yielded, which?

```html
<!DOCTYPE this-will-be-quirks-mode>
<div class="SaMe sAmE SAME same">
```

---

If we perform deduplication and yield in quirks mode, it doesn't seem like there's any way to get it right. It really feels to me like we should just yield all the unique class names and the fact that repeated class names are yielded is something that calling code can deal with if they desire.

This will only affect quirks mode, which is expected to be the less common mode. The Tag Processor and HTML (Fragment) Proccessor will operate in no-quirks mode by default.

Reply:
> It really feels to me like we should just yield all the unique class names and the fact that repeated class names are yielded is something that calling code can deal with if they desire.

this is the only situation I truly feel strongly about, and it's that we shouldn't do it, because of problems like the fundamental question: "how many class names are here?" we would be making it imperative for calling code to perform CSS-level semantic parsing when interacting with our CSS functions.

it's possible we're overthinking it. I see value in lower-casing them for normalization purposes (people will likely be checking with lower-case variations in their calling code anyway, even though `has_class()` is here for checking for the presence of a class name).

the reason for reporting the case as-is, for the first class name listed, is that in a way it preserves the author intent. in these situations we have to realize that the markup is invalid in some way. how much do we fix invalid code? well, we can't, but it's our call how to present these situations.

in `set_attribute()` we preserve the casing given for the attribute. maybe this should only be a principle for _setting_ functions and let _getting_ functions read the normalized form (which is how `get_attribute()`) works. in fact, `get_attribute_names_with_prefix()` lower-cases attribute names, so maybe the best option is matching that behavior, where `class_list()` in quirks-mode reports lower-cased variants, but `has_class()` always works, and `add_class()` preserves the provided casing.

Reply:
Ok, I've updated `class_list` to yield lowercased and appropriately de-duplicated class names in quirks mode.